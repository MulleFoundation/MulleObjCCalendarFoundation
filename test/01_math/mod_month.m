#include <stdio.h>
#include <assert.h>


static int  mod_0_n( int *p_value, int max)
{
   int   diff;
   int   value;

   diff  = 0;
   value = *p_value;

   if( value < 0)
   {
      diff  = value / max - 1;
      value %= max;
      if( value)
         value += max;
   }
   else
      if( value >= max)
      {
         diff   = value / max;
         value %= max;
      }

   *p_value = value;
   return( diff);
}


static int  mod_1_13( int *value)
{
   int   diff;

   *value -= 1;
   diff    = mod_0_n( value, 12);
   *value += 1;

   assert( *value >= 1 && *value <= 12);

   return( diff);
}

int  main( void)
{
   int   i;
   int   value;
   int   diff;

   for( i = -25; i <= 25; i++)
   {
      value = i;
      diff  = mod_1_13( &value);
      printf( "%d: mod=%d diff=%d\n", i, value, diff);
   }
   return( 0);
}

