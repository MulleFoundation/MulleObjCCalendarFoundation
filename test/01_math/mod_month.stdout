-25: mod=11 diff=-3
-24: mod=12 diff=-3
-23: mod=1 diff=-3
-22: mod=2 diff=-2
-21: mod=3 diff=-2
-20: mod=4 diff=-2
-19: mod=5 diff=-2
-18: mod=6 diff=-2
-17: mod=7 diff=-2
-16: mod=8 diff=-2
-15: mod=9 diff=-2
-14: mod=10 diff=-2
-13: mod=11 diff=-2
-12: mod=12 diff=-2
-11: mod=1 diff=-2
-10: mod=2 diff=-1
-9: mod=3 diff=-1
-8: mod=4 diff=-1
-7: mod=5 diff=-1
-6: mod=6 diff=-1
-5: mod=7 diff=-1
-4: mod=8 diff=-1
-3: mod=9 diff=-1
-2: mod=10 diff=-1
-1: mod=11 diff=-1
0: mod=12 diff=-1
1: mod=1 diff=0
2: mod=2 diff=0
3: mod=3 diff=0
4: mod=4 diff=0
5: mod=5 diff=0
6: mod=6 diff=0
7: mod=7 diff=0
8: mod=8 diff=0
9: mod=9 diff=0
10: mod=10 diff=0
11: mod=11 diff=0
12: mod=12 diff=0
13: mod=1 diff=1
14: mod=2 diff=1
15: mod=3 diff=1
16: mod=4 diff=1
17: mod=5 diff=1
18: mod=6 diff=1
19: mod=7 diff=1
20: mod=8 diff=1
21: mod=9 diff=1
22: mod=10 diff=1
23: mod=11 diff=1
24: mod=12 diff=1
25: mod=1 diff=2
